app/dao/fact_billing_dao.py

```diff
@@ -684,3 +684,21 @@ def fetch_usage_year_for_organisation(organisation_id, year):
         service_with_usage[str(email_usage.service_id)]['emails_sent'] = email_usage.emails_sent
 
     return service_with_usage
+
+
+def fetch_billing_details_for_all_services():
+    billing_details = db.session.query(
```

Comment:
I think we can simplify this query, since we're already doing a join to just get the whole service and organisation - it's not a big deal memory-wise, since we're only talking about a few thousand objects. Then we could just have:

```
db.session.query(Service).options(db.joinedload(Service.organisation))
```

so that in the code we can just do:

```
if service.purchase_order_number:
   ... (use service.purchase_order_number)
  else:
   ... (use service.organisation.purchase_order_number)
```

We could even DRY it up more with this technique, noting the column names are the same in both cases, by doing something like:

```
billing_details = service if service.purchase_order_number else organisation
```

Reply:
I like how short and sweet the query would become. Do you think it would also be performant if we had 30 000 services?

Reply:
Yes, it should be. In my experience, there are two important concerns with DB queries:

- Number of records, as an order of magnitude. Eventually this will exceed available memory and we will need to batch the query. Asking for a subset of columns just buys us a bit more time.

- Query complexity (e.g. inefficient indices). We're not doing anything fancy here, though.

You make a good point about the future performance. Since this is only an internal feature, and the number of services grows slowly, I don't think it's worth the effort to optimise this just yet.

So I think it's safe to make the change I'm suggesting.

What do you think?

Reply:
> I don't think it's worth the effort to optimise this just yet

But the effort has been made already, no? It would actually be extra effort to make it shorter but slightly less performant 🤔 

I like the aesthetics of what you propose, but kinda struggle justifying potentially running into problems a bit quicker in the future so that we could have a bit simpler code now.

I think what I would like most is a third opinion from someone like @servingUpAces . That would put my mind at peace :)

Reply:
Sure, a third opinion might be easiest 👍.

I think we should be careful about performance: normally it's better to start with readable code, and only optimise it when we have data to show this is necessary. That's because optimised code is often less readable. Without the data, it's hard to know what to optimise, and by how much - so called "premature optimisation".

I might be missing something, though: do we know there's a performance issue already?

Reply:
No, we do not know of any performance issues here at the moment, and I guess this is an operation we only perform sporadically.

Reply:
Ahh, that is true, I guess we could get away with not checking if organisation exists if we filtered to only get live services, as all live services have orgs and only live services have bills to pay.

Still argument remains for restricting queries to return only what we need to save resources. That does sound more eco-friendly.

Reply:
Hmm...I can see the argument both ways. Thanks for your thoughts @servingUpAces. I'm not sure about _always_ restricting queries, as that would make all our code a lot more verbose.

On the other hand, I can see the advantage of being explicit in this case: it's not obvious that the two tables both have columns with exactly the same names, so can be used interchangeably for billing. So maybe it's better to pull out the columns explicitly, so that someone in the future doesn't need to figure that out.

I'm coming round to leaving it as-is 👍.

> Otherwise you'd have to check that organisation exists before calling service.organisation.purchase order.

That's a good spot!

So we'd need a new conditional for the case where a service didn't have an organisation or its own billing details.

Reply:
Actually I forgot to mention you could use the coalesce function in the select.
`func.coalesce(Service.purchase_order_number, Organisation.purchase_order_number).label('purchase_order_number'),`

Reply:
Oooh, that would be a great compromise!

If we did that, could we get rid of the service vs. org conditional?

Reply:
That is a cool trick, but not sure it would work for the requirements from the ticket:
```
In order to determine where to fetch these values from, first check if there is a 'purchase order number' present for the service. If there is then set these fields using the values stored against the service.
If there is no purchase order number on the service, then check if there is a purchase order number on the parent organisation. If there is then set these fields using the values stored against the organisation.
If neither is present leave those fields empty in the report.
```

Reply:
I asked @yahoopete about this, so let's wait and see.

Reply:
@CrystalPea how would it not match the behaviour described in the ticket?

I had a look and it looks like it should behave the same.

Reply:
Ah yes, I see what you mean. Not quite the utopia I imagine.

(Arguably this makes it the worst option, as the behaviour could be unpredictable 🙈 . Oh well, at least now we're all really familiar with this code!)